src/hooks/useAsync.ts

```diff
@@ -17,6 +16,11 @@ export type UseAsyncReturnType<Result, Error = unknown> =
       loading: false;
       result: null;
       error: Error;
+    }
+  | {
+      loading: false;
+      result: null;
+      error: null;
```

Comment:
Is this case to indicate that a request hasn't been made yet? If it is, we probably only want this case in `useAsyncOnDemand`'s return type. We may also want to add a `status` property to indicate whether the request has gone out or if it's re-requesting or something. Maybe like this:
```ts
export type UseAsyncOnDemandReturnType<Result, Error = unknown> = 
| {
      loading: true;
      result: null;
      error: null;
      status: "IN_FLIGHT";
    }
  | {
      loading: false;
      result: Result;
      error: null;
      status: "DONE"
    }
  | {
      loading: false;
      result: null;
      error: Error;
      status: "DONE"
    }
  | {
      loading: false;
      result: null;
      error: null;
      status: "UNSENT";
    }
```

It doesn't have to be exactly like that, but we'll want to be able to tell the difference between a request that returns `null` (which is valid JSON) and a request that hasn't been sent yet.